File: game/effects/splat.py
import random

import pygame

from config import Config
from tween import Tween


class Splat:
    splats = None
    orb = None

    @staticmethod
    def init():
        Splat.splats = [pygame.image.load("resources/splats/splat0" + str(idx) + ".png").convert_alpha() for idx in range(10)]
        Splat.orb = pygame.image.load("resources/orbs/orb.png").convert_alpha()

    @staticmethod
    def render_splat(screen, time):
        surface = Splat.splats[random.randint(0, 9)]

        if time > Config.EFFECT_DURATION_SPLAT / 2:
            alpha = int(Tween.map(time, Config.EFFECT_DURATION_SPLAT / 2, Config.EFFECT_DURATION_SPLAT, 255, 0))
            surface.set_alpha(alpha)
        else:
            surface.set_alpha(255)

        screen.blit(surface, (0, 0))

File: game-web/src/effects/splat.ts
/** Port of game/effects/splat.py */

import { Config } from '../config';
import { Tween } from '../tween';

async function fetchImage(url: string): Promise<ImageBitmap> {
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`Splat: HTTP ${resp.status} for ${url}`);
  const blob = await resp.blob();
  try {
    return await createImageBitmap(blob);
  } catch (e) {
    throw new Error(`Splat: createImageBitmap failed for ${url} — ${e}`);
  }
}

export class Splat {
  static splats: ImageBitmap[] = [];
  static orb: ImageBitmap | null = null;

  static async init(): Promise<void> {
    const base = import.meta.env.BASE_URL;
    const splatPromises = Array.from({ length: 10 }, (_, i) =>
      fetchImage(`${base}splats/splat0${i}.png`),
    );
    Splat.splats = await Promise.all(splatPromises);
    Splat.orb = await fetchImage(`${base}orbs/orb.png`);
  }

  static renderSplat(ctx: OffscreenCanvasRenderingContext2D, time: number): void {
    const surface = Splat.splats[Math.floor(Math.random() * 10)];

    let alpha = 1;
    if (time > Config.EFFECT_DURATION_SPLAT / 2) {
      alpha = Tween.map(time, Config.EFFECT_DURATION_SPLAT / 2, Config.EFFECT_DURATION_SPLAT, 1, 0);
      alpha = Math.max(0, Math.min(1, alpha));
    }

    ctx.save();
    ctx.globalAlpha = alpha;
    ctx.drawImage(surface, 0, 0);
    ctx.restore();
  }
}
